### ISSUE TITLE
Few comments

### ISSUE THREAD

**Issue by skeydan** [kasaai/cvae-claims#17]
Just a write-up of what came to mind inspecting the model:

1. 

 Input information might  get very dominated by the output of the `grus` (which all are parallel so possibly compute the same thing?), perhaps we lose the other types of information here?:

```
> cvae$cond
Tensor("concatenate/concat:0", shape=(?, 291), dtype=float32)
```

In general, perhaps try if encoding that information in MLP units would perform the same or better (as anyway we're not returning sequences)?

2.
`num_categories` seems to be a bit magic for me, it doesn't surface anywhere as it gets introduced and then flattened into `self$draw_input_flattened`...

3. 

Just a question - I didn't extract that piece of code but I didn't understand why we need `tfd_independent(mix, reinterpreted_batch_ndims = NULL)`

4. 
Also I don't know if with the `repeat_vector` in the middle, the model still has enough information to produce the timesteps afresh?

5. 
As the most general point, I wonder if it makes sense testing simpler components of the architecture separately:
- e.g., use TFP in a non-CVAE setting (MLP/RNN)?
- use CVAE without TFP?

6. 
As another idea for testing, perhaps create a maximally-easy-to-learn dataset and see if the architecture works there?

**Comment by kevinykuo** [comment]
Thanks! Here are the responses and follow up questions:

1. We're using GRU for the sequential input because the data is sequential and not fixed length (we use masking in the RNN to work around that). What do you mean by "parallel" in "which all are parallel so possibly compute the same thing?"

2. We're using a continuous relaxation of the (multidimensional) one-hot categorical distribution for the latent. To use the distribution we have to specify what it looks like, kinda like specifying the dimensionality in the typical Gaussian case.

3. Since we're returning a sequence in the decoder, if we didn't have the `tfd_independent()` call, we'd have 11 time steps of a random variable, and it'd be harder to extract the `log_prob`s for the reconstruction loss.

4. What would be a good alternative here? I took inspiration from https://blog.keras.io/building-autoencoders-in-keras.html

```
inputs = Input(shape=(timesteps, input_dim))
encoded = LSTM(latent_dim)(inputs)

decoded = RepeatVector(timesteps)(encoded)
decoded = LSTM(input_dim, return_sequences=True)(decoded)
```

5 & 6. Modeling explicit density seems to be the most natural for the business problem since we want to be able to sample different future realizations, so it's either going to be CVAE or some form of conditional autoregressive flow. I think the components are behaving as expected and I believe we are learning, since the model is good at putting the mode of the data (0s). Is the goal of trying simpler components to make sure the building blocks are working correctly? Either way, I'll strip down the predictors and see how we do with simpler models.

**Comment by skeydan** [comment]
1) I didn't express that well, better forget about it ;-) Perhaps one could compare with a (padded) MLP but then, perhaps it's not worth the effort...

4) I've never experimented with autoencoders for seq. data myself, so I don't really have experience when that works and when not ... We had a few github issues where my impression was it might not be perfect, but then, perhaps it's totally adequate for the data in case...

5/6) Yeah again I might be overly skeptical but I like isolating blocks to see if they work as I think... especially since sometimes, I've found it hard to  "replicate" others' results even when I saw their code
